### DISCUSSION TITLE
Custom source controllers

### DISCUSSION THREAD

**Discussion by stv0g** [fluxcd/flux2#4849] [Q&A]
I am currently trying to build a new controller which consumes events from the `source-controller` to get notified about new commits in a `GitRepository`.

This new controller fetches the latest artifact from the `GitRepository` and then performs some transformation/evaluation of the repository contents.

I would like to make these evaluated contents available as a source so it can be consumed by the other controllers in the GOTK (e.g. `kustomize-controller` or `helm-controller`).

### Background

I am using a [Nix](https://nixos.org/) Flake to declare my Kubernetes resources.
My custom controller is running `nix build` to evaluate my Nix expressions which return a the usual K8S & Flux resources.

### Possible Approaches

1. I could push to the evaluated resources to another `GitRepository` or another branch of the source repo
2. I could use another source such as an `OCIRepository` or `Bucket`
3. I could try to fake a `GitRepository`, and manually update its `.status.artifact` attribute to point to my controller.
4. Maybe I could create a new CRD for my own source?

The last point brings me to my main question:

**Does `the kustomize-controller` and/or `helm-controller` support custom source CRDs?**

### References

In this video a similar approach is presented using ArgoCD and its hooks:

https://media.ccc.de/v/nixcon-2023-35290-nix-and-kubernetes-deployments-done-right

Also interesting:

- https://github.com/mayflower/argocd-nix-flakes-plugin

**Comment by stv0g** [comment]
This could maybe also realized via kustomize plugins. However, they are not available in Flux due to security reasons:

- https://github.com/fluxcd/kustomize-controller/issues/323

But nevertheless, I think some sort of pre-processing a source must be possible?

**Comment by souleb** [comment]
`Kustomize-controller` accepts `OCIRepository;GitRepository;Bucket` as sources and `Helm-controller` `OCIRepository;HelmChart`.
Using an `OCIRepository` source would enable consumptions from both.
